include/lbann/proto/proto_helpers.hpp

```diff
@@ -0,0 +1,62 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_PROTO_PROTO_HELPERS_HPP_INCLUDED
+#define LBANN_PROTO_PROTO_HELPERS_HPP_INCLUDED
+
+#include <google/protobuf/message.h>
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace lbann
+{
+namespace proto
+{
+
+template <typename OutT, typename... Args>
+struct GenerateBuilderType_struct
+{
+  using type = std::function<std::unique_ptr<OutT>(Args...)>;
+};
+
+template <typename OutT, typename... Args>
+using generate_builder_type =
+  typename GenerateBuilderType_struct<OutT, Args...>::type;
```

Comment:
The idea is that you can partially specialize a struct but not a `using` statement. If you have some `OutT` class whose objects can be constructed by some object other than `std::function`, you could specialize the `struct` for that type while keeping the `generate_builder_type` "clean" (since its entire purpose is to get rid of the `typename ...::type`stuff). At this point it doesn't matter (there's no overhead to this); if you feel strongly, I can remove it.

Reply:
Ah, I see. That sounds reasonable.